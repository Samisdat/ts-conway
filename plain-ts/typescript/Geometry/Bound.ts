export class Bound {

    private min: number;

    private max: number;

    constructor(min: number, max: number) {

        if (max < min) {
            throw new Error('max can not be lower then min');
        }

        this.min = min;
        this.max = max;
    }

    public isWithin(test: number): boolean {
        return (test >= this.min && test <= this.max);
    }

    public isBelow(test: number): boolean {
        return (test < this.min);
    }

    public isAbove(test: number): boolean {
        return (test > this.max);
    }

    public confine(value: number): number {

        if (value < this.min) {
            return this.min;
        }

        if (value > this.max) {
            return this.max;
        }

        return value;

    }

    public getMax(): number{
        return this.max;
    }

    public getMin(): number{
        return this.min;
    }

    public expand(value: number): void{

        if(true === this.isWithin(value)){
            return;
        }

        if(true === this.isBelow(value)){
            this.min = value
        }

        if(true === this.isAbove(value)){
            this.max = value
        }

    }

}
