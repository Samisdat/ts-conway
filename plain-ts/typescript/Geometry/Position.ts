export class Position {

    public readonly  x: number;
    public readonly  y: number;

    constructor(x: number, y: number) {

        if (-0 === x) {
            x = 0;
        }

        if (-0 === y) {
            y = 0;
        }

        this.x = x;
        this.y = y;
    }

    public move(position: Position): Position {
        return new Position(
            this.x + position.x,
            this.y + position.y
        );
    }

    public inverse(): Position {
        return new Position(
            -1 * this.x,
            -1 * this.y
        );
    }

    public compare(position: Position): boolean {

        return (position.x === this.x && position.y === this.y);

    }

    public getNeighbours(): Position[] {

        const neighbours: Position[] = [];

        // neighbours on bottom
        neighbours.push(
            new Position(this.x - 1, this.y - 1)
        );
        neighbours.push(
            new Position(this.x, this.y - 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y - 1)
        );

        // neighbours on the left and on the right
        neighbours.push(
            new Position(this.x - 1, this.y)
        );
        neighbours.push(
            new Position(this.x + 1, this.y)
        );

        // neighbours on top
        neighbours.push(
            new Position(this.x - 1, this.y + 1)
        );
        neighbours.push(
            new Position(this.x, this.y + 1)
        );
        neighbours.push(
            new Position(this.x + 1, this.y + 1)
        );

        return neighbours;
    }

    public clone(): Position {
        return new Position(
            this.x,
            this.y
        );
    }

    public toString(): string {

        let str = '';

        str += (0 <= this.x) ? '+' : '';
        str += this.x;

        str += '/';

        str += (0 <= this.y) ? '+' : '';
        str += this.y;

        return str;

    }
}
