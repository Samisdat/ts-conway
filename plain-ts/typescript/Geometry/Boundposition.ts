import {Position} from '@Conway/Geometry/Position';
import {Bound} from '@Conway/Geometry/Bound';

export class Boundposition {

    private x: Bound;
    private y: Bound;

    constructor(topLeft: Position, bottomRight: Position) {

        if (bottomRight.x < topLeft.x) {
            throw new Error('bottomRight must be right from topLeft');
        }

        if (bottomRight.y < topLeft.y) {
            throw new Error('bottomRight must be below from topLeft');
        }

        this.x = new Bound(topLeft.x, bottomRight.x);
        this.y = new Bound(topLeft.y, bottomRight.y);

    }

    public isWithin(test: Position): boolean {
        return (this.x.isWithin(test.x) && this.y.isWithin(test.y));
    }

    public confine(value: Position): Position {

        return new Position(
            this.x.confine(value.x),
            this.y.confine(value.y)
        );

    }

    public bottomLeft(): Position{

        return new Position(this.x.getMin(), this.y.getMin());

    }

    public topRight(): Position{

        return new Position(this.x.getMax(), this.y.getMax());

    }

    public getWidth(): number{

        return this.x.getMax() - this.x.getMin() + 1;

    }

    public getHeight(): number{

        return this.y.getMax() - this.y.getMin() + 1;

    }

    public expand(position: Position): void{

        if(true === this.isWithin(position)){
            return;
        }

        this.x.expand(position.x);
        this.y.expand(position.y);

    }

}
