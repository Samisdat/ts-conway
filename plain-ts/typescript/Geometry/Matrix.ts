import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

interface MatrixInterface {
    [index: string]: Position;
}


export class Matrix {

    private matrix: MatrixInterface = {};

    private boundposition: Boundposition;

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;


        if(undefined === this.boundposition){
            this.boundposition = new Boundposition(position, position);
        }
        else{
            this.boundposition.expand(position);
        }

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public getAll(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    public getBound(): Boundposition {

        return this.boundposition;

    }

    public getWidth(): number {

        return this.getBound().getWidth();

    }

    public getHeight(): number {

        return this.getBound().getHeight();

    }

}
