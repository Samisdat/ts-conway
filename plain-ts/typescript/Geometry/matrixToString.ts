import {Matrix} from '@Conway/Geometry/Matrix';
import {Position} from '@Conway/Geometry/Position';

interface NumberedScale {
    sign: string[];
    number: string[];
};

const getSigns = (numberedScale: NumberedScale, i: number) =>{
    if(0 === i % 5){

        numberedScale.number.push(Math.abs(i).toString());

        if(0 > i){
            numberedScale.sign.push('-');
        }
        else if(0 < i){
            numberedScale.sign.push('+');
        }
        else{
            numberedScale.sign.push(' ');
        }

    }
    else{
        numberedScale.number.push(' ');
        numberedScale.sign.push(' ');
    }

};

const getScaleHorizontal = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const left = matrix.getBound().bottomLeft().x;
    const right  = matrix.getBound().topRight().x;

    matrixArray = matrixArray.reverse();

    const newRow: string[] = []
    for(let i = 0; i < matrixArray[0].length; i += 1){
        newRow.push('')
    }

    matrixArray.push(newRow);

    const scale: NumberedScale = {
        sign: [],
        number: [],
    };

    for(let i = left; i <= right; i += 1){

        getSigns(scale, i);
    }

    matrixArray.push(scale.number);
    matrixArray.push(scale.sign);

    matrixArray = matrixArray.reverse();
    return matrixArray;


};

const getScaleVertical = (matrix: Matrix, matrixArray: string[][]): string[][] => {

    const top  = matrix.getBound().topRight().y;
    const bottom = matrix.getBound().bottomLeft().y;

    const scale: NumberedScale = {
        sign: [' ', ' ', ' '],
        number: [' ', ' ', ' '],
    };

    for(let i = top; i >= bottom; i -= 1){

        getSigns(scale, i);

    }

    for(let i = 0; i < matrixArray.length; i += 1){

        matrixArray[i] = matrixArray[i].reverse();

        matrixArray[i].push(' ')
        matrixArray[i].push(' ')

        matrixArray[i].push(scale.number[i])
        matrixArray[i].push(scale.sign[i])

        matrixArray[i] = matrixArray[i].reverse();

    }

    return matrixArray;

};

const addScale = (matrix: Matrix, matrixString: string[][]): string[][] => {

    matrixString = getScaleHorizontal(matrix, matrixString);
    matrixString = getScaleVertical(matrix, matrixString);
    return matrixString;
};

export const matrixToString = (matrix: Matrix, scale = true): string => {

    if(0 === matrix.getAll().length){
        return 'Matrix is empty';
    }

    const width = matrix.getWidth();
    const height = matrix.getHeight();

    const topLeft = new Position(
        matrix.getBound().bottomLeft().x,
        matrix.getBound().topRight().y
    );

    let matrixOfStrings: string[][] = [];

    let pointer = topLeft.clone();

    for(let y = 0; y < height; y += 1){

        const line = [];

        for(let x = 0; x < width; x += 1){

            if(true === matrix.has(pointer)){
                line.push('O');
            }
            else{
                line.push('.');
            }

            pointer = pointer.move(new Position(1,0));

        }

        matrixOfStrings.push(line);

        pointer = new Position(
            topLeft.x,
            pointer.y - 1
        );

    }

    if(true === scale){
        matrixOfStrings = addScale(matrix, matrixOfStrings);
    }

    return matrixOfStrings.map((row)=>{return row.join('')}).join('\n');;

}
