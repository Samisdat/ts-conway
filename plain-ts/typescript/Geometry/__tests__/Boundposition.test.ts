import {Position} from '@Conway/Geometry/Position';
import {Boundposition} from '@Conway/Geometry/Boundposition';

describe('PositionBound', () => {

    test('can be created', () => {

        const boundposition = new Boundposition(
            new Position(-2, -2),
            new Position(2, 2)
        );

        expect(boundposition).toBeInstanceOf(Boundposition);

    });

    test('create fails when right is not right from left ', () => {

        const createBound = function () {
            new Boundposition(
                new Position(2, 0),
                new Position(-2, 0)
            );
        };

        expect(createBound).toThrowErrorMatchingSnapshot();

    });

    test('create fails when bottom is not below from top', () => {

        const createBound = function () {
            new Boundposition(
                new Position(1, 2),
                new Position(2, 1)
            );
        };

        expect(createBound).toThrowErrorMatchingSnapshot();

    });

    test('within', () => {

        const boundposition = new Boundposition(
            new Position(-2, -2),
            new Position(2, 2)
        );

        expect(boundposition.isWithin(new Position(-3, -3))).toBeFalsy();
        expect(boundposition.isWithin(new Position(-2, -3))).toBeFalsy();
        expect(boundposition.isWithin(new Position(-3, -2))).toBeFalsy();

        expect(boundposition.isWithin(new Position(-2, -2))).toBeTruthy();
        expect(boundposition.isWithin(new Position(-1, -1))).toBeTruthy();
        expect(boundposition.isWithin(new Position(0, 0))).toBeTruthy();
        expect(boundposition.isWithin(new Position(1, 1))).toBeTruthy();
        expect(boundposition.isWithin(new Position(2, 2))).toBeTruthy();

        expect(boundposition.isWithin(new Position(3, 2))).toBeFalsy();
        expect(boundposition.isWithin(new Position(2, 3))).toBeFalsy();
        expect(boundposition.isWithin(new Position(3, 3))).toBeFalsy();

    });

    test('confine', () => {

        const boundposition = new Boundposition(
            new Position(-2, -2),
            new Position(2, 2)
        );

        expect(boundposition.confine(new Position(-3, -3))).toStrictEqual(new Position(-2, -2));
        expect(boundposition.confine(new Position(-2, -3))).toStrictEqual(new Position(-2, -2));
        expect(boundposition.confine(new Position(-3, -2))).toStrictEqual(new Position(-2, -2));

        expect(boundposition.confine(new Position(-2, -2))).toStrictEqual(new Position(-2, -2));
        expect(boundposition.confine(new Position(-1, -1))).toStrictEqual(new Position(-1, -1));
        expect(boundposition.confine(new Position(0, 0))).toStrictEqual(new Position(0, 0));
        expect(boundposition.confine(new Position(1, 1))).toStrictEqual(new Position(1, 1));
        expect(boundposition.confine(new Position(2, 2))).toStrictEqual(new Position(2, 2));

        expect(boundposition.confine(new Position(3, 2))).toStrictEqual(new Position(2, 2));
        expect(boundposition.confine(new Position(2, 3))).toStrictEqual(new Position(2, 2));
        expect(boundposition.confine(new Position(3, 3))).toStrictEqual(new Position(2, 2));

    });

    test('expand not needed, while already withing', () => {

        const boundposition = new Boundposition(
            new Position(-3, -3),
            new Position(2, 2)
        );

        expect(boundposition.bottomLeft()).toStrictEqual(new Position(-3, -3));
        expect(boundposition.topRight()).toStrictEqual(new Position(2, 2));

        const within = new Position(0,0);

        boundposition.expand(within);

        expect(boundposition.bottomLeft()).toStrictEqual(new Position(-3, -3));
        expect(boundposition.topRight()).toStrictEqual(new Position(2, 2));

    });

    test('expand leftBottom', () => {

        const boundposition = new Boundposition(
            new Position(-3, -3),
            new Position(2, 2)
        );

        expect(boundposition.bottomLeft()).toStrictEqual(new Position(-3, -3));
        expect(boundposition.topRight()).toStrictEqual(new Position(2, 2));

        const within = new Position(-10,-11);

        boundposition.expand(within);

        expect(boundposition.bottomLeft()).toStrictEqual(new Position(-10, -11));
        expect(boundposition.topRight()).toStrictEqual(new Position(2, 2));

    });

    test('width -> both x above 0', () => {

        const boundposition = new Boundposition(
            new Position(2, 2),
            new Position(3, 2)
        );

        expect(boundposition.getWidth()).toBe(2);

    });

    test('width -> both x below 0', () => {

        const boundposition = new Boundposition(
            new Position(-3, 2),
            new Position(-2, 2)
        );

        expect(boundposition.getWidth()).toBe(2);

    });

    test('width -> one 0 and one above', () => {

        const boundposition = new Boundposition(
            new Position(0, 2),
            new Position(2, 2)
        );

        expect(boundposition.getWidth()).toBe(3);

    });

    test('width', () => {

        const boundposition = new Boundposition(
            new Position(1, 2),
            new Position(3, 2)
        );

        expect(boundposition.getWidth()).toBe(3);

    });

    test('height', () => {

        const boundposition = new Boundposition(
            new Position(1, 2),
            new Position(3, 2)
        );

        expect(boundposition.getHeight()).toBe(1);

    });

    test('height', () => {

        const boundposition = new Boundposition(
            new Position(1, -1),
            new Position(3, 1)
        );

        expect(boundposition.getHeight()).toBe(3);

    });

    test('height', () => {

        const boundposition = new Boundposition(
            new Position(1, 0),
            new Position(3, 2)
        );

        expect(boundposition.getHeight()).toBe(3);

    });

    test('height', () => {

        const boundposition = new Boundposition(
            new Position(1, 1),
            new Position(3, 3)
        );

        expect(boundposition.getHeight()).toBe(3);

    });

});
