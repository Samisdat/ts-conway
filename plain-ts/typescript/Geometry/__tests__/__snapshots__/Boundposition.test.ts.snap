// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionBound create fails when bottom is not below from top 1`] = `"bottomRight must be below from topLeft"`;

exports[`PositionBound create fails when right is not right from left  1`] = `"bottomRight must be right from topLeft"`;
